libs/telemetry/src/main/java/org/opensearch/telemetry/tracing/SpanFactory.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.telemetry.tracing;
+
+import org.opensearch.telemetry.tracing.noop.NoopSpan;
+
+import java.util.function.Supplier;
+
+/**
+ * Factory to create spans based on the configured level
+ */
+public final class SpanFactory {
+
+    private final TracingTelemetry tracingTelemetry;
+
+    private final Supplier<Level> levelSupplier;
+
+    /**
+     * Creates SpanFactor instance with provided level supplier and tracing telemetry
+     * @param levelSupplier configured level supplier
+     * @param tracingTelemetry tracing telemetry
+     */
+    public SpanFactory(Supplier<Level> levelSupplier, TracingTelemetry tracingTelemetry) {
+        this.levelSupplier = levelSupplier;
+        this.tracingTelemetry = tracingTelemetry;
+    }
+
+    /**
+     * Creates span with provided arguments
+     * @param spanName name of the span
+     * @param parentSpan span's parent span
+     * @param level of the span
+     * @return span instance
+     */
+    public Span createSpan(String spanName, Span parentSpan, Level level) {
+        return isLevelEnabled(level) ? createDefaultSpan(spanName, parentSpan, level) : createNoopSpan(spanName, parentSpan, level);
+    }
+
+    private boolean isLevelEnabled(Level level) {
+        Level configuredLevel = levelSupplier.get();
+        return level.isHigherOrEqual(configuredLevel);
+    }
+
+    private Span createDefaultSpan(String spanName, Span parentSpan, Level level) {
+        Span telemetrySpan = tracingTelemetry.createSpan(spanName, getLastValidSpanInChain(parentSpan), level);
+        return telemetrySpan;
+    }
+
+    private NoopSpan createNoopSpan(String spanName, Span parentSpan, Level level) {
+        return new NoopSpan(spanName, parentSpan, level);
+    }
+
+    private Span getLastValidSpanInChain(Span parentSpan) {
```

Comment:
I think this is very confusing behaviour to replace span's parents: no only it is surprising, it makes insanely difficult to compare the same trace but with different levels (let say info and debug).

Reply:
In this method we are actually ignoring the no-op spans, which got created because the corresponding span levels were less than the configured trace level. More details in [this](https://github.com/opensearch-project/OpenSearch/pull/7648/#discussion_r1233273938) comment

Reply:
I understand what it does, my point is that it changes the view of the spans hierarchy.

Reply:
Sorry but isn't it ok? The spans which are getting removed/overridden from spans hierarchy are no-op spans. What would you recommend instead?

Reply:
> Sorry but isn't it ok? 

My take on it - no, it is not ok. Let's me have an example for you, we are tracing the **same** flow (does not matter which one), the first is INFO level, the second if DEBUG level. I should be able to compare both traces (since those are for the same flow) to see what is the difference. This task becomes literally impossible when hierarchy is changed in arbitrary ways.

> What would you recommend instead?

There are two options I see here:
a) drop levels from spans, at this moment I think we don't have a need for them and they in add the complexity 
b) ensure that child spans are not using level higher that parent spans, it will keep hierarchy unchanged

Reply:
> a) drop levels from spans, at this moment I think we don't have a need for them and they just add the complexity

I know you might know this but just re-iterating to ensure we are on the same page. The idea behind introducing level concept in span was to dynamically alter the tracing depth (similar to logging). The code flow could emit multiple spans (instrumented with different levels). With level in the span, we would have the ability to choose which one to activate and which one to mute in runtime. Without it either all the spans would get activated or none. How do you suggest we achieve this behavior?

> b) ensure that child spans are not using level higher that parent spans, it will keep hierarchy unchanged

We had thought of this approach earlier. But we moved to current approach because we thought this approach would constraint the user. For example the user would have to always be aware of the level of the parent span to ensure her/his current span is not a no-op. However with the current approach, she/he doesnt have remember the parent's span level. She/He could just instrument the code based on her/his understanding of the importance(/criticality) of the span and the framework will take care of rest. Love to hear your thoughts on this though.

Reply:
> The idea behind introducing level concept in span was to dynamically alter the tracing depth (similar to logging).

That's the thing, traces are not logs, they are inherently hierarchical, vs flat logs.  Level for logs just means no log record, level for trace spans (as it implemented) means arbitrary hierarchical transformation.

> She/He could just instrument the code based on her/his understanding of the importance(/criticality) of the span and the framework will take care of rest. Love to hear your thoughts on this though.

I think I have expressed  my thoughts before - the idea of level does not cleanly apply to hierarchical traces (I have never seen any distributed tracing framework that does that). Please notice that user in this case are OpenSearch contributors and plugin / extension authors - we are not building generic tracing framework.

Reply:
I think we can get rid of levels altogether. Ideally, if we are in agreement to support "operation", which would be an OpenSearch entity to control span limits, then it should just be a simple switch to enable/disable the tracing on the basis of operation. For example, we can enable/disable operation - search, indexing etc dynamically. 
OpenTelemetry controls the volume of span/attributes using span limits - https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#span-limits. We should explore something similar which we can set globally or for individual operation. Also, sampling should be another way to control volume of spans generated if we don't want to enable/disable complete operation. Similar to what opentelemetry supports - https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#sampler `TraceIdRatioBased `.

Reply:
Adding to why we introduced Level and think its needed - With level we would be able to deep-dive into code path and  identify parts(/spans) which is taking more time. This would be helpful while debugging latency issues. By default we could keep system level setting at say TERSE/INFO level to limit the number of spans, as each span creation would have some overhead. During debugging we could lower tracing level to see more spans.

Hence I think we should have level support. But I also agree with your point of changing hierarchy. I have addressed it in latest commit by ensuring the child's level is always less than parent's level. @reta Please have a look when you get chance.

Reply:
@rishabhmaurya I believe Operation is another dimension we need for (horizontally)controlling the trace output. We would need Level for (vertically)controlling the depth of the trace. More details in above comment. The alternatives you mentioned(limits/sampling etc.) helps in limiting the trace output not controlling them as they dont add/remove spans with much contextual logic. We would though need them anyways just for limiting use-cases.

Reply:
>  I have addressed it in latest commit by ensuring the child's level is always less than parent's level. @reta Please have a look when you get chance.

Thanks @rishabhmaurya , I think if we keep levels, this is they way to manage them.

> With level we would be able to deep-dive into code path and identify parts(/spans) which is taking more time. This would be helpful while debugging latency issues.

The levels do not help with debugging latency issues at, they make it harder - instead of getting the right trace right away, we would need to do some levels acrobatics and wait for the same operation to be invoked so we could capture more spans - certainly not a smooth experience.

> I think we can get rid of levels altogether. Ideally, if we are in agreement to support "operation", which would be an OpenSearch entity to control span limits, then it should just be a simple switch to enable/disable the tracing on the basis of operation. 

I am :100: with @rishabhmaurya here, sampling is the way go, not levels. The one incredibly power advanced sampling techniques used by some companies is adaptive sampling: the decisions to collect the trace (+ all spans) is taken at the end of the trace (not upfront as usually), and one of the decision points is latency - if trace is breaking SLO metrics (which one could only known when it is finished), it is backfilled into the system. The levels would screw that up, not help. 

To conclude,  @suranjay please remove levels altogether, they could be introduced later on when it becomes clear how they should work be used.

Reply:
@reta I have removed the levels from now to unblock this PR as we both are strongly opinionated with contrary thoughts. Will raise another issue to discuss and get alignment on it.